### PR TITLE
Introduce a file_info function.

### PR DESCRIPTION
**Pull request by mhvk**
Which uses a new `.info()` method on binary file readers.

Logically, this would be extended to the stream readers.

### CODE REVIEW

File: baseband/vdif/base.py [hunk 1]
```diff
@@ -62,12 +63,41 @@
...
-class VDIFFileReader(VLBIFileBase):
+class VDIFFileReader(VLBIFileReaderBase):
     """Simple reader for VDIF files.
 
-    Adds ``read_frame`` and ``read_frameset`` methods on top of a binary
-    file reader (which is wrapped as ``self.fh_raw``).
+    Wraps a binary VDIF file, adding convenience methods such as
+    ``read_frame`` and ``get_frame_rate``.
     """
+    def info(self):
```
Comment (cczhu) [inline baseband/vdif/base.py]: Silly idea - should `repr` rely on `info` for information, rather than header0?

File: baseband/dada/base.py [hunk 2]
```diff
@@ -109,6 +109,26 @@ class DADAFileReader(VLBIFileBase):
     Adds a ``read_frame`` method to the basic VLBI binary file wrapper. By
     default, the payload is mapped rather than fully read into physical memory.
     """
+    def info(self):
+        old_offset = self.tell()
```
Comment (cczhu) [inline baseband/dada/base.py]: One-liner docstring (repeat for all others):
```
"""Provides a `dict` of file properties, such as the format and frame rate."""
```

File: baseband/formats.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+
+FILE_FORMATS = {'dada': 'DADA',
+                'mark4': 'Mark4',
+                'mark5b': 'Mark5B',
+                'vdif': 'VDIF',
+                'gsb': 'GSB'}
+
+
+def file_info(name, fmt=None, **kwargs):
```
Comment (cczhu) [inline baseband/formats.py]: Not a huge fan of `fmt`; would prefer `format` (can still use `fmt` in the for loop).

File: baseband/formats.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+    if isinstance(fmt, tuple):
+        for fmt_ in fmt:
+            info = file_info(name, fmt_, **kwargs)
+            if info:
+                break
+
+        return info
+
+    module = importlib.import_module('.' + fmt, package='baseband')
```
Comment (cczhu) [inline baseband/formats.py]: I don't like `module` here, but can't think of something better (`baseband_format` seems excessive, and probably confusing).

File: baseband/gsb/base.py [hunk 5]
```diff
@@ -52,6 +67,16 @@ def write_timestamp(self, header=None, **kwargs):
             header = GSBHeader.fromvalues(**kwargs)
         header.tofile(self.fh_raw)
 
+    def get_frame_rate(self):
+        oldpos = self.tell()
+        self.seek(0)
+        try:
+            timestamp0 = self.read_timestamp()
+            timestamp1 = self.read_timestamp()
+            return 1./(timestamp1.time - timestamp0.time)
```
Comment (cczhu) [inline baseband/gsb/base.py]: Really nitpicking now: add spaces between "/".

File: baseband/dada/base.py [hunk 6]
```diff
@@ -128,6 +148,15 @@ def read_frame(self, memmap=True):
         """
         return DADAFrame.fromfile(self.fh_raw, memmap=memmap)
 
+    def get_frame_rate(self):
```
Comment (cczhu) [inline baseband/dada/base.py]: DOcstring:
```
"""Returns the number of frames per second in a DADA file.

Returns
-------
frame_rate : `~astropy.units.Quantity`
    Frames per second.
"""
```
Though a one-liner would be fine too.

File: baseband/mark5b/base.py [hunk 7]
```diff
@@ -64,6 +92,40 @@ def read_frame(self):
...
+        """Returns the number of frames per second.
+
+        Returns
+        -------
+        frame_rate : `~astropy.units.Quantity`
+            Frames per second.
+
+        Notes
+        -----
+        This function first tries to determine the frame rate by looking for
```
Comment (cczhu) [inline baseband/mark5b/base.py]: "This function" -> "This method" (already corrected under #201, so hoping it makes it to here once we merge).

File: baseband/mark5b/base.py [hunk 8]
```diff
@@ -64,6 +92,40 @@ def read_frame(self):
                                     ref_time=self.ref_time, nchan=self.nchan,
                                     bps=self.bps)
 
+    def get_frame_rate(self):
+        """Returns the number of frames per second.
```
Comment (cczhu) [inline baseband/mark5b/base.py]: `"""...frames per second in a Mark 5B file."""`

File: baseband/mark5b/base.py [hunk 9]
```diff
@@ -64,6 +92,40 @@ def read_frame(self):
...
+            return super(Mark5BFileReader, self).get_frame_rate()
+        except Exception as exc:
+            oldpos = self.tell()
+            self.seek(0)
+            try:
+                header0 = self.read_header()
+                self.seek(header0.payload_nbytes, 1)
+                header1 = self.read_header()
+                tdelta = header1.fraction - header0.fraction
+                assert tdelta != 0.
```
Comment (cczhu) [inline baseband/mark5b/base.py]: Add a useful error message about this ("First two headers have identical timestamps, possibly because the file's data rate is above 512 Mbps.  Try explicitly passing in a sample rate.")?  I think it gets passed back to the stream reader.  Should also have a test.

(Obviously this is a strange edge case - when would we get a < 1 second file with that high a sampling rate?)

File: baseband/vlbi_base/base.py [hunk 10]
```diff
@@ -46,6 +46,44 @@ def __repr__(self):
...
+        """Returns the number of frames per second.
+
+        Returns
+        -------
+        frame_rate : `~astropy.units.Quantity`
+            Frames per second.
+
+        Notes
+        -----
+        The function cycles through headers, starting from the file start, to
```
Comment (cczhu) [inline baseband/vlbi_base/base.py]: function -> method (again, corrected already by #201)

File: baseband/mark5b/base.py (hunk 8 above)
Comment (mhvk) [reply baseband/mark5b/base.py]: I think that should be obvious if one opened one!

**Comment by mhvk** [comment]
Thanks for the comments.  I'll implement those for #201; for `info()`, the main goal really was to show what could be done.

**Comment by mhvk** [comment]
I rather drastically changed this, but think it is now much nicer.
```
In [1]: from baseband.data import *; from baseband import file_info

In [2]: file_info(SAMPLE_MARK4)
Out[2]: 
format = mark4
frame_rate = 400.0 Hz
sample_rate = 32.0 MHz
samples_per_frame = 80000
sample_shape = (8,)
bps = 2
complex_data = False
offset0 = 2696

missing:  decade, ref_time: needed to infer full times.

In [3]: file_info(SAMPLE_MARK4, decade=2010)
Out[3]: 
format = mark4
frame_rate = 400.0 Hz
sample_rate = 32.0 MHz
samples_per_frame = 80000
sample_shape = SampleShape(nchan=8)
bps = 2
complex_data = False
start_time = 2014:167:07:38:12.47500
ntrack = 64
offset0 = 2696
stop_time = 2014:167:07:38:12.48000
size = 1280000
```

A general opener is nearly trivial with this...

File: baseband/core.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+                'gsb': 'GSB'}
+
+
+def file_info(name, format=None, **kwargs):
+    """Get format and other information from a baseband file.
+
+    Parameters
+    ----------
+    name : str or filehandle
+        Raw file for which the information is to be gotten.
```
[the remark this answers is not in the document]
Comment (mhvk) [reply baseband/core.py]: Done

File: baseband/core.py [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+        (i.e., ``info()``).  Any extra keywords used in opening the file are
+        attached as an ``kwargs`` attribute.
+
+    Notes
+    -----
+    The keyword arguments will only be used if needed, so if one is unsure
+    what format a file is, but knows it was taken recently and has 8 channels,
+    one would put in ``ref_time=Time('2015-01-01'), nchan=8``. Alternatively,
+    and perhaps easier, one can first call the function without extra arguments
+    in which case the result will describe what is missing.
```
[the remark this answers is not in the document]
Comment (mhvk) [reply baseband/core.py]: I agree it should be both places; one wants `file_info?` to be useful...

So, let's do docs after...

File: baseband/gsb/file_info.py [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+        # Part of our information, including the format, comes from the
+        # underlying timestamp file.
+        self._fh_ts_info = self._parent.fh_ts.info
+        self._fh_ts_info_attrs = self._fh_ts_info.attr_names
+        extra_attrs = tuple(attr for attr in self._parent_attrs
+                            if attr not in self._fh_ts_info_attrs)
+        self.attr_names = self._fh_ts_info_attrs + extra_attrs
+
+    def _up_to_date(self):
+        # Stream readers cannot after initialization, so the check is easy.
```
[the remark this answers is not in the document]
Comment (mhvk) [reply baseband/gsb/file_info.py]: rephrased

File: baseband/vlbi_base/file_info.py [hunk 14]
```diff
@@ -0,0 +1,179 @@
+# Licensed under the GPLv3 - see LICENSE
+"""Provide a base class for "info" properties.
+
+Loosely based on astropy.utils.data_info.DataInfo
```
[the remark this answers is not in the document]
Comment (mhvk) [reply baseband/vlbi_base/file_info.py]: done

File: baseband/vlbi_base/file_info.py [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+        # We link to attributes from the parent rather than just overriding
+        # __getattr__ to allow us to look for changes.
+        for attr in self._parent_attrs:
+            setattr(self, attr, getattr(self._parent, attr))
+        self.missing = {}
+
+    def _up_to_date(self):
+        """Determine whether the information we have stored is up to date."""
+        return all(getattr(self, attr) == getattr(self._parent, attr)
+                   for attr in self._parent_attrs)
```
[the remark this answers is not in the document]
Comment (mhvk) [reply baseband/vlbi_base/file_info.py]: I don't think it is possible, since other properties that are determined may depend on these parent propeties; for instance, think of `Mark4FileReader` where someone updates `decade` after initialization - now `start_time` suddenly becomes available. I'll add a comment to the code to make clear this is why it is done this way.

File: baseband/vlbi_base/file_info.py [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+            return None
+        finally:
+            fh.seek(old_offset)
+
+    def _get_format(self):
+        return self._parent.__class__.__name__.split('File')[0].lower()
+
+    def _get_start_time(self):
+        try:
+            return self.header0.time
```
[the remark this answers is not in the document]
Comment (mhvk) [reply baseband/vlbi_base/file_info.py]: Ah, yes, I should override `_get_start_time` and ensure the `sample_rate` has been calculated before that is called. Similarly for Mark 5B, where one might need the `frame_rate`.

[the remark this answers is not in the document]
Comment (mhvk) [reply baseband/vlbi_base/file_info.py]: This also suggests I should have been less lackadaisical about testing in the individual formats...

**Comment by mhvk** [comment]
Just pushed the changes so far; to do is to add additional tests (I quite like finding out that an `.m5a` file is in fact VDIF!)

**Comment by mhvk** [comment]
OK, now also added tests to ensure the fix actually works. I still feel it would be better to have a bit more exhaustive tests (basically, a `test_file_info` for every format), but that would not have caught the problem you found, so perhaps the spot checks for problematic cases are in fact more useful.

**Comment by mhvk** [comment]
Added more tests, changes the `repr` yet again, but also made sure that the info atttributes are all in the docstring one sees when doing `fh.info?` as well as in the documentation.

If things pass, I'll go ahead and override the review, since I think the comments have been addressed.
